trie-db/src/trie_codec.rs

```diff
@@ -390,6 +390,17 @@ pub fn decode_compact<L, DB, T>(db: &mut DB, encoded: &[Vec<u8>])
 	where
 		L: TrieLayout,
 		DB: HashDB<L::Hash, T>,
+{
+	decode_compact_from_iter::<L, DB, T, _>(db, &mut encoded.iter().map(Vec::as_slice))
+}
+
+/// Variant of 'decode_compact' that accept an iterator of encoded nodes as input.
+pub fn decode_compact_from_iter<'a, L, DB, T, I>(db: &mut DB, encoded: &'a mut I)
+	-> Result<(TrieHash<L>, usize), TrieHash<L>, CError<L>>
+	where
+		L: TrieLayout,
+		DB: HashDB<L::Hash, T>,
+		I: Iterator<Item = &'a [u8]>,
```

Comment:
Is it a common pattern to accept a mutable reference to an iterator? I've never seen before, do we use it somewhere?
I'd expect to take an iterator by value. Or `IntoIterator`.
Something like this should work:
```suggestion
pub fn decode_compact_from_iter<'a, L, DB, T, I>(db: &mut DB, encoded: I)
	-> Result<(TrieHash<L>, usize), TrieHash<L>, CError<L>>
	where
		L: TrieLayout,
		DB: HashDB<L::Hash, T>,
		I: IntoIterator<Item = &'a [u8]>,
```

Reply:
My first implementation did use Iterator directly but then there is some need of awkward sytax for 'a, and &mut makes it easy to attach the lifetime.
And just now I realize something interesting with &mut: you can use the iterator again.
Decode compact does not have to read all the nodes, and return the number of read nodes, the original intention was to have multiple trie encoded in the input.
With &mut you can simply use the iterator again to get next trie (in the code i am writing I read top trie then iterate on it then for each child trie root in the proof I decode trie from the remaining nodes in the proof).

Reply:
If I pass the iterator as parameter then I will need for every child trie in proof to build a new iterator and then use seek to skip the n previous elements.
But does '&mut Iterator' implement intoiterator and can be reuse, I need to check that :)